woocommerce/woocommerce-android
# Add Analytics Usage Tracking

Part of #5756 and 162-gh-woocommerce/woomobile-private.

## Background

In p91TBi-6Cl-p2, we discussed how we can track the “Analytics/My Store tab usage”. A single tap event does not necessarily mean that they're using it. We settled on Josh Heald's suggestion (p91TBi-6Cl-p2#comment-7391) where we use a combination of time and number of interactions to determine usage. 

## Changes

This was already implemented on iOS (https://github.com/woocommerce/woocommerce-ios/pull/5503 and https://github.com/woocommerce/woocommerce-ios/pull/5820). I added the same thing for Android. I pretty much just tried to translate the Swift code to Kotlin. 😅 

There is one main class, [`MyStoreStatsUsageTracksEventEmitter`](https://github.com/woocommerce/woocommerce-android/blob/issue/5756-add-analytics-usage-tracking/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsUsageTracksEventEmitter.kt), that receives “interacted” signals from the various parts of the Analytics UI. This class calculates how many interactions and how long they were apart from each other. It will only track the `USED_ANALYTICS` event if all these conditions pass:

- They interacted with the UI for at least 10 seconds
- They interacted with the UI (taps, scrolls, etc) for at least 5 times
- They did not idle for at least 20 seconds in between interactions

All of these are configured in:

https://github.com/woocommerce/woocommerce-android/blob/217f94bd9ca690de3f103f8de9a63f197368b118/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsUsageTracksEventEmitter.kt#L36-L54

These values should match the iOS implementation, [`StoreStatsUsageTracksEventEmitter.swift`](https://github.com/woocommerce/woocommerce-ios/blob/61284460ddbf66ccf2dfa5c6a1bcd87c15e145e1/WooCommerce/Classes/ViewRelated/Dashboard/Stats%20v4/StoreStatsUsageTracksEventEmitter.swift#L9-L28).

The UI should call `MyStoreStatsUsageTracksEventEmitter.interact()` function when the user performed these actions:

- Scrolling 🚧 
- Pull-to-refresh
- Tapping on the bars in the chart
- Changing the tab
- Navigating to the My Store tab
- Tapping on a product in the Top Performers list

### Not Yet Implemented: Scrolling

I have not added scrolling yet. I wanted to call `interact()` only when we detect that the user **started** scrolling. But it doesn't look like [`NestedScrollView` ](https://developer.android.com/reference/androidx/core/widget/NestedScrollView) has an event that provides this. It only has [`OnScrollChangeListener`](https://developer.android.com/reference/androidx/core/widget/NestedScrollView.OnScrollChangeListener) which is triggered way too many times to be useable. 

I'll tackle this separately. Suggestions welcome! **Update**: Created https://github.com/woocommerce/woocommerce-android/pull/5965

Sliding on the chart also triggers many events but I don't think it's that terrible compared to the number of events triggered by scrolling. iOS also behaves similarly on the chart. 

## Testing

1. Log in
2. Navigate to My Store
3. Do a few of the following:
     - Pull-to-refresh
     - Tapping on the bars in the chart
     - Changing the tab
     - Navigating to the My Store tab
     - Tapping on a product in the Top Performers list
4. Confirm that after performing at least 5 of the interactions within more than 10 seconds, a `used_analytics` Tracks event is logged.

## Reviewing

Only 1 reviewer is needed but anyone can review. 

📢 Please nitpick me to the death. My skills are not up to date so I would love any suggestion about how to make my code more idiomatic and better. 🙃 

## Author Checklist
